src/SqlBinding/SqlAsyncCollector.cs

```diff
@@ -90,55 +92,284 @@ public async Task FlushAsync(CancellationToken cancellationToken = default)
         /// <param name="rows"> The rows to be upserted </param>
         /// <param name="attribute"> Contains the name of the table to be modified and SQL connection information </param>
         /// <param name="configuration"> Used to build up the connection </param>
-        private async Task UpsertRowsAsync(string rows, SqlAttribute attribute, IConfiguration configuration)
+        private async Task UpsertRowsAsync(IEnumerable<T> rows, SqlAttribute attribute, IConfiguration configuration)
         {
             using (SqlConnection connection = SqlBindingUtilities.BuildConnection(attribute.ConnectionStringSetting, configuration))
             {
-                string tableName = SqlBindingUtilities.NormalizeTableName(attribute.CommandText);
+                string fullDatabaseAndTableName = attribute.CommandText;
+                ObjectCache cachedTables = MemoryCache.Default;
+                TableInformation tableInfo = cachedTables[fullDatabaseAndTableName] as TableInformation;
+
+                if (tableInfo == null)
+                {
+                    Console.WriteLine("lookin");
+                    tableInfo = await TableInformation.RetrieveTableInformationAsync(connection, fullDatabaseAndTableName);
+
+                    // If we were unable to look up the primary keys, for whatever reason, return early. 
+                    // We'll try again next time. TODO: do we need to have a max # of retries / backoff plan?
+                    if (tableInfo == null)
+                    {
+                        return;
```

Comment:
Would this look like a silent failure to the customer? Would an exception be a better experience?

Reply:
We do throw exceptions in `RetrieveTableInformationAsync` if we run into any issues, so this line of code is probably dead code.